### ISSUE TITLE
build_tools/llvm_version.txt vs `llvm-project` as a submodule?

### ISSUE THREAD

**Issue by rolfmorel** [plaidml/tpp-mlir#936]
I was a bit surprised by the instructions in the [current README](https://github.com/plaidml/tpp-mlir/blob/f7ea8fd05acbafbbba3ff840cd4fa16cc8c5124f/README.md) regarding checking out the designated `llvm-project` version.

The instructions are to:
```
# Clone
git clone https://github.com/llvm/llvm-project.git

# checking out a tpp-mlir compatible version of llvm-project
wget https://raw.githubusercontent.com/plaidml/tpp-mlir/main/build_tools/llvm_version.txt
pushd llvm-project
git checkout `cat ../llvm_version.txt`
popd
rm llvm_version.txt
```

To me this seems to reproduce what `git submodule`s are good for though does so manually. As I currently don't see advantages to this approach, my suggestions are as follows:

1. Switch to just using a submodule to keep track of the designated commit.
2. Add a short document `build_tools/llvm_version.readme.txt` documenting why the current approach is preferable over using submodules.

Either solution would have stopped me from making this issue. Other takes on this are of course appreciated as well.

**Comment by rengolin** [comment]
The main reason we use `llvm_version.txt` is that we build it only once per commit, install, and then all tpp-mlir builds in the CI re-use the same shared version. Not only users logged in in the cluster can reuse the same install, but all builders, too. 

We have a lot of builders and they're all building in different directories. By using sub modules, even using `ccache` we'd still be building LLVM on every builder every time we bump it.

Another reason why we do that is that we have support for building tpp-mlir as an _out-of-tree_ LLVM project. If we had it as a sub module too, we'd create a cycle. Sure, CMake can probably hack it around, but we like to avoid using too advanced CMake options. :)

Another issue with sub modules is that it makes it hard to develop multiple branches in LLVM itself. We do a lot of upstreaming, and we want to test multiple choices of LLVM with tpp-mlir. If we use sub modules, we have to keep branches and build again and again every time we switch. But since we already have an LLVM tree elsewhere, using worktrees, with separate build directories, etc. we can just switch between them and just re-build tpp-mlir (which is much quicker).

All in all, sub modules are convenient when you don't change the code and the build time for sub modules are a fraction of the build time of your project. In our case, none of that is true.

**Comment by rengolin** [comment]
Added a quick note on building multiple LLVM trees in parallel and using it for multiple sub-projects on the same file system at https://github.com/plaidml/tpp-mlir/wiki/Multiple-LLVM-Workspaces#building
